### ISSUE TITLE
In System Tray, cannot open Network, Sound and Power flyout.

### ISSUE THREAD

**Issue by dragonfirecmv** [cairoshell/cairoshell#50]
The icon responsible for showing the status of each entity does not respond to any click.

**Comment by dremin** [comment]
I'm exploring a few different possibilities here:
1. Find a way to make these work
1. Hide them, and instead rely on a button to open Action Center
1. Implementing these directly into Cairo.

The problem is they are owned by explorer.exe, which we are trying to replace 😄 I'll keep this issue updated, but in the meantime I highly recommend [Volume2](https://irzyxa.wordpress.com) for volume control; it's incredibly flexible and is what I have been using.

**Comment by dragonfirecmv** [comment]
Actually, I have an idea and here's my proposal:

Since I've been doing a custom implementation of macOS shells too, I would suggest adding custom execution code for these special entity of Windows 10 tray icon by custom linking them into their corresponding shell commands, such as 'ms-availablenetworks:' for Networks flyout, using system key combinations for Action Center, etc.

**Comment by dragonfirecmv** [comment]
Whoops, accidentally closed the issue!

**Comment by dremin** [comment]
Thanks for reminding me of the available networks URI! For Action Center, I have implemented the `ms-actioncenter:` URI. It looks like I could use `sndvol -f` to display a volume slider, albeit the old one. Unfortunately, I haven't found a way to activate the modern Volume or Power flyouts yet.

**Comment by dragonfirecmv** [comment]
I think I got a better idea: https://github.com/cairoshell/cairoshell/issues/54

**Comment by dremin** [comment]
I have started this by adding Action Center support in the latest Milestone 3 release. Please check for updates.

**Comment by dragonfirecmv** [comment]
Thanks! Just the flyout issues left then.

**Comment by mdmk** [comment]
As a workaround for lack of sound dropdown, you can use EarTrumpet if you are on Windows 10.
It's a clone of the default tray icon, but with a more convenient volume mixer.
But I encounter a crash when lauched after Cairo, of after a few click if lauched before Cairo. I fill in a separate issue for this bug.

**Comment by mdmk** [comment]
As seen in #77, try disabling "Periodically re-establish notification area hook".
Ear Trumpet and native network tray will work correctly then.

**Comment by dremin** [comment]
One alternative network icon solution is to change the action of clicking the network icon. You can do this via a registry change on Windows 10:

HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Control Panel\Settings\Network
1. Right-click the registry key > Permissions > Change Owner to yourself and then give yourself full control
2. Double-click "ReplaceVan"
3. Use one of the following settings:
0: Default
1: Open Network settings
2: Windows 8 behavior

**Comment by cosmicdan** [comment]
You might of already discovered this, but in Windows 10 (and 8 I assume) the explorer shell is built like a UWP app. The flyouts and Action Center and all that are using Xaml rather than GDI+ - so it can't be manipulated easily.

The explorer shell used to be an app appearing in the WindowsApps folder in earlier builds, but it's since been moved somewhere secret. In my research, I think the only way to truly figure out how to hook/simulate this behavior is by finding where the explorer shell UI stuff is and reverse engineering it.

**Comment by dragonfirecmv** [comment]
@cosmicdan correct. However in previous builds Windows create an entry point (usually a shell command or keyboard shortcut) that triggers them, and while it isn't a perfect implementation, it serves the purpose. It seems that Microsoft 'forgot' to reimplement them at the moment.

**Comment by cosmicdan** [comment]
Indeed, Microsoft "forgot" to implement a lot of things in Windows 10's explorer :)

I'll keep an eye on this thread in case somebody finds a solution. I too am working on a shell ~~replacement~~ augmentation (not something that completely replaces the environment like Cairo does but just adds a Gnome-like toolbar to the top Window) so I would be interested in any discovery anybody might make in this area.

Good to see how far Cairo shell has come, though - glad it's still in active development!

**Comment by dremin** [comment]
I'm not sure the problem is that the flyouts are XAML per se-- here's the issue as I see it (based on my time on the issue, but who knows, maybe I'm way far off):

Explorer's taskbar window (Shell_TrayWnd) is still present when Cairo is open due to all of the features Explorer provides. Cairo gets notification area icons by creating its own Shell_TrayWnd and advertising that a new taskbar was created. Some icons do some positioning and other logic based on the location of Shell_TrayWnd. Unfortunately sometimes Cairo's window is found first, sometimes Explorer's is found first. If Explorer's window is found first, the XAML flyout for Network works just fine even when invoked from Cairo. However, if Cairo's window is found first, for some reason the flyout doesn't open.

This is why the Network flyout works if the "periodically re-establish notification area hook" option is disabled in Cairo settings. All this option does when enabled is re-create and re-advertise a new Shell_TrayWnd every 10 seconds or so (yes, it's nasty), which causes Cairo's Shell_TrayWnd to be found before Explorer's usually. So, disabling this option means that Explorer's Shell_TrayWnd is found first.

To solve this we need to figure out what the flyouts are looking for that Cairo isn't providing an expected result to.

**Comment by dremin** [comment]
The Network flyout in Windows 10 should be fully functional now [in this test build](https://ci.appveyor.com/api/buildjobs/co9l8knpxu1iggp1/artifacts/Installer%2FCairoSetup_64bit.exe) **without** needing to disable the "periodically re-establish notification area hook" option. Could someone please double-check on their system as well? Thanks!

**Comment by josuave** [comment]
Network fly out is working but shows up in the same position it would if the windows taskbar is visible.

Sent from my iPhone

On Mar 27, 2018, at 7:31 PM, Sam Johnson <notifications@github.com<mailto:notifications@github.com>> wrote:


The Network flyout in Windows 10 should be fully functional now in this test build<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fci.appveyor.com%2Fapi%2Fbuildjobs%2Fco9l8knpxu1iggp1%2Fartifacts%2FInstaller%252FCairoSetup_64bit.exe&data=02%7C01%7C%7Cc6f21b7ba16d49045fbc08d5943ae26b%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636577902990007341&sdata=FzXeLXYkMPPMXzGr9j9%2FYtTOOVzz84bkpEEX8CgiSCk%3D&reserved=0> without needing to disable the "periodically re-establish notification area hook" option. Could someone please double-check on their system as well? Thanks!

**Comment by dremin** [comment]
The working network flyout is in the latest release! Is anyone able to try the Power flyout?

**Comment by mdmk** [comment]
Network flyout working, but icon closes and opens panel.
On first click it opens, if you click further it will close quickly and open.
Didn't try behavior on escape and click outside.

Power flyout not working but right click does (it is a "old" contextual menu, not a black modern one like Network flyout)

Sound flyout not working as before.

**Comment by dremin** [comment]
Sound is available in the latest build.

**Comment by josuave** [comment]
A trick to allow the Network and other Explorer notify apps (WiFi) to display in the upper right quadrant.
Quit Cairo, unlock windows taskbar if needed, dock it to the top edge, restart cario. I am investigating a way to do this in code. It seems windows 10 broke moving the taskbar via API

**Comment by dremin** [comment]
By chance have recent improvements enabled the Power icon to work?

**Comment by josuave** [comment]
Worked for me. I have also gotten in the habit of moving my windows taskbar to the top of the screen prior to running Cairo to ensure that the shell hosted tray apps appear in the correct location

Sent from my iPhone

On Apr 27, 2018, at 8:17 PM, Sam Johnson <notifications@github.com<mailto:notifications@github.com>> wrote:


By chance have recent improvements enabled the Power icon to work?

**Comment by skyborgff** [comment]
@dremin Did you end up finding the ms-actioncenter: URI But for the modern Sound and Power?

**Comment by dremin** [comment]
@skyborgff I ended up implementing sound another way, and heard that power is working correctly in the latest builds. Are these working for you?

**Comment by skyborgff** [comment]
@dremin I'm not using this project, i wanted to know for a project of my own (a rainmeter skin). could you provide me the power URI (if it exists) please?

**Comment by dremin** [comment]
@skyborgff Unfortunately I did not find a URI for power.
